tinygrad/renderer/cstyle.py

```diff
@@ -297,17 +294,16 @@ class HIPLanguage(CStyleLanguage):
 __device__ void vstore_half2(float2 data, size_t offset, half *p) { *(p + offset*2) = (half)data.x; *(p + offset*2 + 1) = (half)data.y; }
 __device__ void vstore_half4(float4 data, size_t offset, half *p) {
   *(p + offset*4) = (half)data.x; *(p + offset*4 + 1) = (half)data.y; *(p + offset*4 + 2) = (half)data.z; *(p + offset*4 + 3) = (half)data.w; }
-__device__ half exp2(half x) { return hexp2(x); }
-__device__ half log2(half x) { return hlog2(x); }
-__device__ half sin(half x) { return hsin(x); }
-__device__ half sqrt(half x) { return hsqrt(x); }
-__device__ half hmax(half a, half b) { return __hgt(a, b) ? a : b; }
-__device__ half operator%(const half &a, const half &b) { return __hsub(a, __hmul(b, __float2half(floorf(__half2float(a) / __half2float(b))))); }
   """
   gid = [f'blockIdx.{chr(120+i)}' for i in range(3)]
   lid = [f'threadIdx.{chr(120+i)}' for i in range(3)]
   xid = [f'(blockIdx.{chr(120+i)}*blockDim.{chr(120+i)}+threadIdx.{chr(120+i)})' for i in range(3)]
-  code_for_op = {**CStyleLanguage().code_for_op, BinaryOps.MAX: lambda a,b,dtype: f"max({a},{b})" if dtype != dtypes.half else f"hmax({a},{b})",}
+  code_for_op = {
+    **CStyleLanguage().code_for_op, **{op: lambda a,dtype,op=op: f"h{CStyleLanguage().code_for_op[op]('(half)'+a,dtype)}" if dtype == dtypes.half else f"float4({CStyleLanguage().code_for_op[op](a+'.x',dtype)}, {CStyleLanguage().code_for_op[op](a+'.y',dtype)}, {CStyleLanguage().code_for_op[op](a+'.z',dtype)}, {CStyleLanguage().code_for_op[op](a+'.w',dtype)})" if dtype == dtypes.float.vec(4) else f"{CStyleLanguage().code_for_op[op](a,dtype)}" for op in UnaryOps}, # noqa: E501
+    UnaryOps.NEG: lambda a,dtype: f"(-(half){a})" if dtype == dtypes.half else CStyleLanguage().code_for_op[UnaryOps.NEG](a,dtype), # noqa: E501
+    **{op: lambda a, b, dtype, op=op, fxn=fxn: f"__h{fxn}((half){a},(half){b})" if dtype == dtypes.half else CStyleLanguage().code_for_op[op](a,b,dtype) for op, fxn in zip([BinaryOps.ADD, BinaryOps.SUB, BinaryOps.MUL, BinaryOps.DIV], ["add","sub","mul","div"])}, # noqa: E501
+    BinaryOps.MAX: lambda a,b,dtype: f"__hgt((half){a},(half){b})?(half){a}:(half){b}" if dtype == dtypes.half else f"float4(max({a}.x,{b}.x),max({a}.y,{b}.y),max({a}.z,{b}.z),max({a}.w,{b}.w))" if dtype == dtypes.float.vec(4) else f"max({a},{b})",# noqa: E501
```

Comment:
Careful, this runs things twice with __hgt.

And why do you need float4 here? max shouldn't be done on float4